src/Host/Monitor/Impl/CredentialManager.cs

```diff
@@ -0,0 +1,434 @@
+                                logger?.LogError(Resources.Error_LogOnAttemptFailed, errorCode);
+                                if (token != IntPtr.Zero) {
+                                    CloseHandle(token);
+                                }
+                                continue;
+                            }
+
+                            // Credentials received successfully.
+                            // Exit loop, another attempt is not needed.
+                            break;
+                        } else {
+                            // CredUIPromptForWindowsCredentials failed to load some component
+                            logger?.LogError(Resources.Error_CredUIFailedToLoad, Marshal.GetLastWin32Error());
+                            throw new Win32Exception(string.Format(Resources.Error_CredUIFailedToLoad, ret));
+                        }
+                    }
+                } finally {
+                    if (token != IntPtr.Zero) {
+                        CloseHandle(token);
+                    }
+
+                    if (credentialBuffer != IntPtr.Zero) {
+                        Marshal.ZeroFreeCoTaskMemUnicode(credentialBuffer);
+                    }
+                }
+
+                return attempts;
+            }
+        }
+
+        public static void RemoveCredentials(ILogger logger = null) {
+            CredDelete(BrokerUserCredName, CredType.GENERIC, 0);
+            logger?.LogInformation(Resources.Info_RemovedCredentials);
+        }
+
+        private static object _setCredentialsOnProcessLock = new object();
+        public static void SetCredentialsOnProcess(ProcessStartInfo psi, ILogger logger = null) {
+            lock (_setCredentialsOnProcessLock) {
+                IntPtr cred = IntPtr.Zero;
+                try {
+                    if (CredRead(BrokerUserCredName, CredType.GENERIC, 0, out cred)) {
+                        Credential credential = new Credential();
+                        credential = (Credential)Marshal.PtrToStructure(cred, typeof(Credential));
+
+                        byte[] passwordBytes = new byte[credential.credentialBlobSize];
+                        Marshal.Copy(credential.credentialBlob, passwordBytes, 0, credential.credentialBlobSize);
+                        string passwordText = Encoding.Unicode.GetString(passwordBytes);
+
+                        var usernameBldr = new StringBuilder(257);
+                        var domainBldr = new StringBuilder(257);
+
+                        uint error = CredUIParseUserName(credential.userName, usernameBldr, usernameBldr.Capacity, domainBldr, domainBldr.Capacity);
+                        if (error != 0) {
+                            logger?.LogError(Resources.Error_UserNameParsing, error);
+                            throw new Win32Exception((int)error);
+                        }
+
+                        SecureString pass = new SecureString();
```

Comment:
ToSecureString(), but it is a last resort. Better send native straight to SecureString as above